onnxruntime/core/graph/graph.cc

```diff
@@ -4213,67 +4250,159 @@ const ONNX_NAMESPACE::GraphProto& Graph::ToGraphProto() {
+  if (!GraphProtoSyncNeeded()) {
+    result = *graph_proto_;
+    ORT_THROW_IF_ERROR(ProcessSubgraphsInMemoryData(result, /*process_main*/ true));
+  } else {
+    ToGraphProtoInternal(result);
 
-      // set in raw data
-      output.clear_data_location();
-      output.set_raw_data(data, tensor_byte_size);
-    }
-  };
+    ORT_THROW_IF_ERROR(ProcessSubgraphsInMemoryData(result, /*process_main*/ false));
 
-  auto* mutable_initializers = result.mutable_initializer();
+    // Add initializers to parent graph by copy converting them from graph_proto_
+    // ToGraphProtoInternal() does not copy initializers for the main graph
+    auto* mutable_initializers = result.mutable_initializer();
 
 #if !defined(DISABLE_SPARSE_TENSORS)
-  const auto& model_path = ModelPath();
-  // We want to make sure that sparse initializers do not appear
-  // as dense duplicates within the initializers list.
-  const bool has_sparse_initializers = !sparse_tensor_names_.empty();
-  const auto sparse_end = sparse_tensor_names_.end();
-  for (const auto& initializer : graph_proto_->initializer()) {
-    if (!has_sparse_initializers || sparse_end == sparse_tensor_names_.find(initializer.name())) {
-      add_initializer(*mutable_initializers, initializer);
-    } else {
-      auto& sparse_initializer = *result.add_sparse_initializer();
-      auto status = utils::DenseTensorToSparseTensorProto(initializer, model_path, sparse_initializer);
-      ORT_ENFORCE(status.IsOK(), "Failed to convert dense initializer to sparse");
+    const auto& model_path = ModelPath();
+    const bool has_sparse_initializers = !sparse_tensor_names_.empty();
+    const auto sparse_end = sparse_tensor_names_.end();
+
+    for (const auto& initializer : graph_proto_->initializer()) {
+      if (!has_sparse_initializers || sparse_end == sparse_tensor_names_.find(initializer.name())) {
+        ORT_THROW_IF_ERROR(InlineOrCopyInitializerToGraphProto(*this, initializer,
+                                                               *mutable_initializers->Add()));
+      } else {
+        auto& sparse_initializer = *result.add_sparse_initializer();
+        if (utils::HasExternalDataInMemory(initializer)) {
+          ONNX_NAMESPACE::TensorProto tensor_proto;
+          ORT_THROW_IF_ERROR(InlineOrCopyInitializerToGraphProto(*this, initializer,
+                                                                 tensor_proto));
+          ORT_THROW_IF_ERROR(utils::DenseTensorToSparseTensorProto(tensor_proto, model_path, sparse_initializer));
+        } else {
+          ORT_THROW_IF_ERROR(utils::DenseTensorToSparseTensorProto(initializer, model_path, sparse_initializer));
+        }
+      }
     }
-  }
 #else
-  for (const auto& initializer : graph_proto_->initializer()) {
-    add_initializer(*mutable_initializers, initializer);
-  }
+    // Add initializers to parent graph by copy converting them from graph_proto_
+    // ToGraphProtoInternal() does not copy initializers for the main graph
```

Comment:
this comment seems redundant since there is a copy of it outside of the ifdef as well.